pkg/apis/v1alpha5/provisioner.go

```diff
@@ -32,63 +34,76 @@ import (
 type ProvisionerSpec struct {
 	// Annotations are applied to every node.
 	//+optional
-	Annotations map[string]string `json:"annotations,omitempty"`
+	Annotations map[string]string `json:"annotations,omitempty" hash:"set"`
 	// Labels are layered with Requirements and applied to every node.
 	//+optional
-	Labels map[string]string `json:"labels,omitempty"`
+	Labels map[string]string `json:"labels,omitempty" hash:"set"`
 	// Taints will be applied to every node launched by the Provisioner. If
 	// specified, the provisioner will not provision nodes for pods that do not
 	// have matching tolerations. Additional taints will be created that match
 	// pod tolerations on a per-node basis.
 	// +optional
-	Taints []v1.Taint `json:"taints,omitempty"`
+	Taints []v1.Taint `json:"taints,omitempty" hash:"set"`
 	// StartupTaints are taints that are applied to nodes upon startup which are expected to be removed automatically
 	// within a short period of time, typically by a DaemonSet that tolerates the taint. These are commonly used by
 	// daemonsets to allow initialization and enforce startup ordering.  StartupTaints are ignored for provisioning
 	// purposes in that pods are not required to tolerate a StartupTaint in order to have nodes provisioned for them.
 	// +optional
-	StartupTaints []v1.Taint `json:"startupTaints,omitempty"`
+	StartupTaints []v1.Taint `json:"startupTaints,omitempty" hash:"set"`
 	// Requirements are layered with Labels and applied to every node.
-	Requirements []v1.NodeSelectorRequirement `json:"requirements,omitempty"`
+	Requirements []v1.NodeSelectorRequirement `json:"requirements,omitempty" hash:"ignore"`
 	// KubeletConfiguration are options passed to the kubelet when provisioning nodes
 	//+optional
-	KubeletConfiguration *KubeletConfiguration `json:"kubeletConfiguration,omitempty"`
+	KubeletConfiguration *KubeletConfiguration `json:"kubeletConfiguration,omitempty" hash:"set"`
 	// Provider contains fields specific to your cloudprovider.
 	// +kubebuilder:pruning:PreserveUnknownFields
-	Provider *Provider `json:"provider,omitempty"`
+	Provider *Provider `json:"provider,omitempty" hash:"ignore"`
 	// ProviderRef is a reference to a dedicated CRD for the chosen provider, that holds
 	// additional configuration options
 	// +optional
-	ProviderRef *MachineTemplateRef `json:"providerRef,omitempty"`
+	ProviderRef *MachineTemplateRef `json:"providerRef,omitempty" hash:"ignore"`
```

Comment:
It seems wrong to me that we would ignore this. We're saying that if this value changes, we don't care? This scares me a little bit since there may be some implicit assumptions about labeling on Nodes for nodeTemplate ownership that might be broken by this. Is there a reason not to drift on this value?

Reply:
Yeah, you're right. We should drift on that value. But we will not drift if the referenced object changes